### DISCUSSION TITLE
Deploying with ClickOnce

### DISCUSSION THREAD

**Discussion by bgoldstein14** [cefsharp/CefSharp#3613] [General]
After deploying with ClickOnce and attempting to bring up the view that uses the cefSharp browser component, the application fails with the error "Could not find file or assembly 'CEFSHARP.CORE.RUNTIME.DLL' or one of its dependencies."

Our project has the main references for cefsharp.wpf, cefsharp.core, and cefsharp as all Copy Local=True. There's nothing else special we're trying to do to get that component to deploy to the desktop.

I've found references to developers in the past having issues with doing this type of deployment but the links are many years old and there doesn't seem to be a concrete solution that ends up being given. One of the solutions points to a non-existent directory to be xcopied during the post-build...but I see how that would help for a ClickOnce deployment anyway.

I see an old discussion (https://github.com/cefsharp/CefSharp/discussions/3402) from here that might have some relevance, but there was no resolution.

Is there a simple list of steps we need to do to get this to deploy and run properly via ClickOnce? My guess is that there's a bunch of file dependencies missing. My other concern with this is that if/when we upgrade the component, this process would have to be repeated otherwise all the dependent files will be invalid.

**Comment by amaitland** [comment]
To better support `ClickOnce/Publish` I've added the option to switch from using the `<None/>` entries to `<Content/>`  this should allow support via a single configuration setting in the `proj` file.

Add the following to a `PropertyGroup` in your `proj` file before the `.targets` entries (which are at the bottom).
```xml
<CefSharpBuildAction>Content</CefSharpBuildAction>
```

**IMPORTANT** :
- Currently only works when `PlatformTarget` is `x86/x64`
- Upgrading users make sure to remove any custom `.targets` files or `ItemGroups` you'd previously used to copy the `CefSharp` dependencies from your `proj` file.

For reference:

- https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-757106286
- https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-791076831

Currently using AnyCPU with clickonce isn't working as detailed in https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-791783346
If you are targeting AnyCPU you can likely self host the browser sub process to workaround the problem.

https://github.com/cefsharp/CefSharp/issues/3407#issuecomment-782828504 for instructions (the code is the same regardless of .Net version).

**Reply by amaitland** [reply]
If you are still having problems then please call CefSharpAfterBuildDiagnostic target after build to get diagnostic output. 

Example at https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj#L144

**Comment by bgoldstein14** [comment]
As it turns out, getting a ClickOnce deployment to work with the component has been extremely tricky. Using various hints throughout this site and various other sites online, I think we've spent a couple hundred hours on trying to get this to work so far.

The next major obstacle that we discovered is the the Subprocess.exe file is being effectively duplicated in the ClickOnce manifest and we don't know how to avoid that, even though we only have it referenced a single time for inclusion into that manifest. I've seen notes on other sites stating that there might be issues referencing a second EXE file in a project (since the first is the project executable itself).

**Do you know how to properly configure the base .csproj to avoid the .exe file being considered a duplicate when msbuild.exe is used to publish the ClickOnce app?**

Thank you.

**Reply by amaitland** [reply]
> If you are still having problems then please call CefSharpAfterBuildDiagnostic target after build to get diagnostic output.

Please call the CefSharpAfterBuildDiagnostic after build and provide a copy of the target output. Example linked above.

**Reply by amaitland** [reply]
Testing with https://github.com/cefsharp/CefSharp.MinimalExample

- Open CefSharp.MinimalExample.sln in VS2019
- Build
- Close Solution
- Edit `CefSharp.MinimalExample.WinForms.csproj` adding `<CefSharpBuildAction>Content</CefSharpBuildAction>` to the first `PropertyGroup` (Example at [CefSharp.MinimalExample.WinForms.csproj#L31](https://github.com/cefsharp/CefSharp.MinimalExample/blob/cefsharp/91/CefSharp.MinimalExample.WinForms/CefSharp.MinimalExample.WinForms.csproj#L31))
- Open CefSharp.MinimalExample.sln in VS2019
- Rebuild
- Comment out `Microsoft.Windows.Common-Controls` from `app.manifest` in `CefSharp.MinimalExample.WinForms`
- Right click `CefSharp.MinimalExample.WinForms` and publish
- Use default options
- Install ClickOnce deployed app.

Loads as expected. Make sure that when installing on client machines that `VC++ 2015 or higher` is installed.

Quick test with converting the project to `PackageReference` from `packages.config`, the resulting installer is missing the `CefSharp.Core.Runtime.dll`, the `.targets` file likely needs a tweak for `PackageReference` scenarios.

**Reply by amaitland** [reply]
If you have `CefSharp` installed in a library project that uses `pacakges.config` things become a little tricky. The [None](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#none) entries will copy the `CEF` resources to a directly referring project, changing to [Content](https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022#content) (changing the `CefSharpBuildAction` to `Content` changes to using `<Content/>` entries) means the files are only copied to the library project, not any referring projects.

1. Installing the `CefSharp` packages in your main project and seeing the adding `<CefSharpBuildAction>Content</CefSharpBuildAction>` to the first `PropertyGroup` of your main project file is one option.
2. Converting your library project to `PackageReference` should allow for the references to flow through to consuming projects as [buildTransitive](https://github.com/NuGet/Home/wiki/Allow-package--authors-to-define-build-assets-transitive-behavior). Noting that I haven't tested this with a mix of `PackageReference` and `packages.config`, so if you mix and match then option `1.` is the fallback if this doesn't work.

**Comment by bgoldstein14** [comment]
Thanks for this additional advice. One other thing to note that complicated things is that we needed to sign the manifest after we did an update to it. It was at that point that the process failed. However, it appears we're finally at a spot where we can move forward with pushing this to production. If we run into more issues, we'll look into the above advice. We might still want to send you a log, but we'll see. We've already spent abundant time on this and have much more to do with the project. 

Thank you again for your timely responses and consideration.

**Comment by amaitland** [comment]
For anyone else that ends up here I've created a more detailed guide for publishing using `ClickOnce` see https://github.com/cefsharp/CefSharp/discussions/4056
